### PR TITLE
[DNM]Rebuild balancer operator steps.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**

### CODE REVIEW

File: server/balancer.go [hunk 1]
```diff
@@ -124,7 +124,14 @@ func (cb *capacityBalancer) selectBalanceRegion(cluster *clusterInfo, stores []*
 	return region, leaderPeer
 }
 
-func (cb *capacityBalancer) selectNewLeaderPeer(cluster *clusterInfo, peers map[uint64]*metapb.Peer) *metapb.Peer {
+func (cb *capacityBalancer) selectNewLeaderPeer(cluster *clusterInfo, peers map[uint64]*metapb.Peer, newPeer *metapb.Peer) *metapb.Peer {
+	// If len(peers) equals to 0, then we cannot find one peer to do leader transfer.
```
Comment (shenli) [inline server/balancer.go]: find one peer -> find any peer

File: server/balancer.go [hunk 2]
```diff
@@ -214,11 +214,18 @@ func (cb *capacityBalancer) Balance(cluster *clusterInfo) (*balanceOperator, err
...
+		log.Warn("new leader peer cannot be found to do balance")
+		return nil, nil
+	}
+
 	addPeerOperator := newAddPeerOperator(newPeer)
+	leaderTransferOperator := newTransferLeaderOperator(oldLeader, newLeader, maxWaitCount)
 	removePeerOperator := newRemovePeerOperator(oldLeader)
 
-	return newBalanceOperator(region, leaderTransferOperator, addPeerOperator, removePeerOperator), nil
+	return newBalanceOperator(region, addPeerOperator, leaderTransferOperator, removePeerOperator), nil
```
Comment (shenli) [inline server/balancer.go]: If newLeader != newPeer, should we put leaderTransferOperator before addPeerOperator?

Comment (siddontang) [inline server/balancer.go]: the newleader is always != newpeer now.

File: server/balancer.go [hunk 3]
```diff
@@ -124,7 +124,14 @@ func (cb *capacityBalancer) selectBalanceRegion(cluster *clusterInfo, stores []*
 	return region, leaderPeer
 }
 
-func (cb *capacityBalancer) selectNewLeaderPeer(cluster *clusterInfo, peers map[uint64]*metapb.Peer) *metapb.Peer {
+func (cb *capacityBalancer) selectNewLeaderPeer(cluster *clusterInfo, peers map[uint64]*metapb.Peer, newPeer *metapb.Peer) *metapb.Peer {
+	// If len(peers) equals to 0, then we cannot find one peer to do leader transfer.
+	// So that we should use new peer.
+	if len(peers) == 0 {
```
Comment (shenli) [inline server/balancer.go]: @siddontang NewLeader could be NewPeer.

Comment (siddontang) [inline server/balancer.go]: I think we should return no peer here.

**Comment by qiuyesuifeng** [comment]
None of use, close it.

**Pull request closed without merging**